Not able to install natively on Apple M1

I'm just trying Owl and couldn't get this to compile natively on a M1 Mac.
If you use the rosetta emulation on the Mac you can get a working system.

I used:
sudo arch -x86_64 before the rest of the "opam/master/shell/install.sh" command. Then opam just works for the full owl install. The other awkward thing is to put arch -x86_64 before all the brew installs eg plplot openblas etc.

I get past AEOS as you say using these settings. The next error is about is about OWL_ARCH_x86_64 & i386, so it looks like setting this to 0 is ignored in the following:
src/owl/core/owl_core_utils.c:255:3: error: invalid output constraint '=a' in asm
# [...]
#   ^
# src/owl/core/owl_macros.h:97:36: note: expanded from macro 'CPUID'
#     __asm__ __volatile__ ("cpuid": "=a" (cpuinfo[0]), "=b" (cpuinfo[1]), "=c" (cpuinfo[2]), "=d" (cpuinfo[3]) : "0" (func), "2" (id) );
#                                    ^
# src/owl/core/owl_core_utils.c:277:5: error: invalid output constraint '=a' in asm
#     CPUID(cpuinfo, 0x0, 0);

Nit: 2nd to last line should be `opam install owl owl-top`

XQuartz? Intense dependencies. It might make sense to pull in something like https://github.com/ocornut/imgui instead.

I died on the last opam install of owl and owl-top:
```
⬇ retrieved eigen.0.3.0  (cached)
[ERROR] The compilation of conf-openblas.0.2.1 failed at "sh -exc cc $CFLAGS $(pkg-config --cflags openblas) test.c $(pkg-config
        --libs openblas)".
[ERROR] The compilation of owl-base.1.0.1 failed at "dune build -p owl-base -j 9".
[ERROR] The compilation of eigen.0.3.0 failed at "dune build -p eigen -j 9".
```
Looks like assembly mnemonic issues and OpenBLAS not getting found, even though it exists at PKG_CONFIG_PATH.

My starting conditions were not clean. I used a separate arch homebrew install. Though, for reference:
```
sudo mkdir -p /usr/local/homebrew
sudo chown `id -u`:`id -g` /usr/local/homebrew
cd /usr/local
git clone https://github.com/Homebrew/brew homebrew
# use the /usr/local/homebrew/bin/brew for this install
arch -x86_64 /usr/local/homebrew/bin/brew install plplot zlib openblas opam
export PKG_CONFIG_PATH="/usr/local/homebrew/opt/openblas/lib/"
opam init   # answer f,y for all rest, made a .zshrc in /usr/local/homebrew
opam install core utop
opam install async yojson core_extended core_bench cohttp async_graphics cryptokit menhir
opam owl owl-top  # breaks here
eval $(opam env)
```

--- a/INSTALL.md
+++ b/INSTALL.md
@@ -15,18 +15,18 @@ before compilation:
 - `OWL_CFLAGS` allows to change the default flags passed to the C targets,
   it defaults to
   ```
-  OWL_CFLAGS="-g -O3 -Ofast -march=native -mfpmath=sse -funroll-loops -ffast-math -DSFMT_MEXP=19937 -msse2 -fno-strict-aliasing -Wno-tautological-constant-out-of-range-compare"`
+  OWL_CFLAGS="-g -O3 -Ofast -funroll-loops -ffast-math -DSFMT_MEXP=19937 -msse2 -fno-strict-aliasing"`
   ```
 
 - `OWL_AEOS_CFLAGS` allows to change the default flags passed to the C targets
   when compiling AEOS. It defaults to
   ```
-  OWL_AEOS_CFLAGS="-g -O3 -Ofast -march=native -funroll-loops -ffast-math -DSFMT_MEXP=19937 -fno-strict-aliasing"
+  OWL_AEOS_CFLAGS="-g -O3 -Ofast -funroll-loops -ffast-math -DSFMT_MEXP=19937 -fno-strict-aliasing"
   ```
 
 - `OWL_DISABLE_LAPACKE_LINKING_FLAG=1` disables the `-llapacke` flag in the
-  linking options. This is useful when you have lapacke installed on
-  non-standard location.
+  linking options. This may be useful when you have certain unusual lapacke
+  setups, do not use it unless you know what you are doing.
 
 If you are not using `opam`, you should run `make clean` before recompiling
 the library after having changed any of those environment variables.

--- a/src/aeos/config/configure.ml
+++ b/src/aeos/config/configure.ml
@@ -92,7 +92,7 @@ let () =
       (* configure link options *)
       let libs = [] @ get_default_libs () @ get_openmp_libs c in
       (* configure compile options *)
-      let cflags = [] @ get_default_cflags @ get_openmp_cflags c in
+      let cflags = [] @ get_default_cflags c @ get_openmp_cflags c in
       (* assemble default config *)
       let conf : C.Pkg_config.package_conf = { cflags; libs } in
       C.Flags.write_sexp "aeos_c_flags.sexp" conf.cflags;

--- a/src/aeos/config/configure.ml
+++ b/src/aeos/config/configure.ml
@@ -83,7 +83,6 @@ let get_default_cflags c =
           ; "__arm__", Switch
           ]
       in
-      let open C in
       match List.map snd defines with
       | Switch true :: _ -> `x86_64
       | _ :: Switch true :: _ -> `x86

--- a/src/aeos/config/configure.ml
+++ b/src/aeos/config/configure.ml
@@ -34,6 +34,7 @@ let get_default_cflags =
   | Not_found ->
     [ "-g"
     ; "-O3"
+    ; "-Ofast"
     ; "-funroll-loops"
     ; "-ffast-math"
     ; "-DSFMT_MEXP=19937"

--- a/src/aeos/config/configure.ml
+++ b/src/aeos/config/configure.ml
@@ -5,6 +5,27 @@
 
 module C = Configurator.V1
 
+let detect_system_header =
+  {|
+  #if __APPLE__
+    #include <TargetConditionals.h>
+    #if TARGET_OS_IPHONE
+      #define PLATFORM_NAME "ios"
+    #else
+      #define PLATFORM_NAME "mac"
+    #endif
+  #elif __linux__
+    #if __ANDROID__
+      #define PLATFORM_NAME "android"
+    #else
+      #define PLATFORM_NAME "linux"
+    #endif
+  #elif WIN32
+    #define PLATFORM_NAME "windows"
+  #endif
+|}
+
+
 let bgetenv v =
   let v' =
     try Sys.getenv v |> int_of_string with
@@ -32,6 +53,25 @@ let get_default_cflags c =
     |> List.filter (fun s -> String.trim s <> "")
   with
   | Not_found ->
+    let os =
+      let header =
+        let file = Filename.temp_file "discover" "os.h" in
+        let fd = open_out file in
+        output_string fd detect_system_header;
+        close_out fd;
+        file
+      in
+      let platform =
+        C.C_define.import c ~includes:[ header ] [ "PLATFORM_NAME", String ]
+      in
+      match List.map snd platform with
+      | [ String "android" ] -> `android
+      | [ String "ios" ]     -> `ios
+      | [ String "linux" ]   -> `linux
+      | [ String "mac" ]     -> `mac
+      | [ String "windows" ] -> `windows
+      | _                    -> `unknown
+    in
     let arch =
       let defines =
         C.C_define.import
@@ -52,10 +92,11 @@ let get_default_cflags c =
       | _ -> `unknown
     in
     [ "-g"; "-O3"; "-Ofast" ]
-    @ (match arch with
-      | `x86_64 -> [ "-march=native"; "-msse2" ]
-      | `arm64  -> [ "-mcpu=apple-m1" ]
-      | _       -> [])
+    @ (match arch, os with
+      | `arm64, `mac -> [ "-mcpu=apple-m1" ]
+      | `arm64, _    -> [ "-march=native" ]
+      | `x86_64, _   -> [ "-march=native"; "-msse2" ]
+      | _            -> [])
     @ [ "-funroll-loops"; "-ffast-math"; "-DSFMT_MEXP=19937"; "-fno-strict-aliasing" ]
 
 
@@ -65,11 +106,9 @@ let get_openmp_cflags c =
   then []
   else (
     match get_os_type c with
-    | "linux"     -> [ "-fopenmp" ]
-    | "linux_elf" -> [ "-fopenmp" ]
-    | "macosx"    -> [ "-Xpreprocessor"; "-fopenmp" ]
-    | "mingw64"   -> [ "-fopenmp" ]
-    | _           -> [])
+    | "mingw64" | "linux" | "linux_elf" -> [ "-fopenmp" ]
+    | "macosx" -> [ "-Xpreprocessor"; "-fopenmp" ]
+    | _ -> [])
 
 
 let get_default_libs () = [ "-lm" ]

--- a/src/aeos/config/configure.ml
+++ b/src/aeos/config/configure.ml
@@ -24,7 +24,7 @@ let get_os_type c =
   | None   -> ""
 
 
-let get_default_cflags =
+let get_default_cflags c =
   try
     Sys.getenv "OWL_AEOS_CFLAGS"
     |> String.trim
@@ -32,14 +32,31 @@ let get_default_cflags =
     |> List.filter (fun s -> String.trim s <> "")
   with
   | Not_found ->
-    [ "-g"
-    ; "-O3"
-    ; "-Ofast"
-    ; "-funroll-loops"
-    ; "-ffast-math"
-    ; "-DSFMT_MEXP=19937"
-    ; "-fno-strict-aliasing"
-    ]
+    let arch =
+      let defines =
+        C.C_define.import
+          c
+          ~includes:[]
+          [ "__x86_64__", Switch
+          ; "__i386__", Switch
+          ; "__aarch64__", Switch
+          ; "__arm__", Switch
+          ]
+      in
+      let open C in
+      match List.map snd defines with
+      | Switch true :: _ -> `x86_64
+      | _ :: Switch true :: _ -> `x86
+      | _ :: _ :: Switch true :: _ -> `arm64
+      | _ :: _ :: _ :: Switch true :: _ -> `arm
+      | _ -> `unknown
+    in
+    [ "-g"; "-O3"; "-Ofast" ]
+    @ (match arch with
+      | `x86_64 -> [ "-march=native"; "-msse2" ]
+      | `arm64  -> [ "-mcpu=apple-m1" ]
+      | _       -> [])
+    @ [ "-funroll-loops"; "-ffast-math"; "-DSFMT_MEXP=19937"; "-fno-strict-aliasing" ]
 
 
 let get_openmp_cflags c =

--- a/src/aeos/config/configure.ml
+++ b/src/aeos/config/configure.ml
@@ -34,8 +34,6 @@ let get_default_cflags =
   | Not_found ->
     [ "-g"
     ; "-O3"
-    ; "-Ofast"
-    ; "-march=native"
     ; "-funroll-loops"
     ; "-ffast-math"
     ; "-DSFMT_MEXP=19937"

--- a/src/owl/config/configure.ml
+++ b/src/owl/config/configure.ml
@@ -76,7 +76,7 @@ let get_ocaml_devmode_flags _c =
   if not enable_devmode then [] else [ "-w"; "-32-27-6-37-3" ]
 
 
-let default_cflags =
+let default_cflags c =
   try
     Sys.getenv "OWL_CFLAGS"
     |> String.trim
@@ -268,7 +268,7 @@ some details on how your openblas has been installed and the output of
         []
         @ openblas_conf.cflags
         @ cblas_conf.cflags
-        @ default_cflags
+        @ default_cflags c
         @ get_devmode_cflags c
         @ get_expmode_cflags c
         @ openmp_config.cflags

--- a/src/owl/config/configure.ml
+++ b/src/owl/config/configure.ml
@@ -5,6 +5,27 @@
 
 module C = Configurator.V1
 
+let detect_system_header =
+  {|
+  #if __APPLE__
+    #include <TargetConditionals.h>
+    #if TARGET_OS_IPHONE
+      #define PLATFORM_NAME "ios"
+    #else
+      #define PLATFORM_NAME "mac"
+    #endif
+  #elif __linux__
+    #if __ANDROID__
+      #define PLATFORM_NAME "android"
+    #else
+      #define PLATFORM_NAME "linux"
+    #endif
+  #elif WIN32
+    #define PLATFORM_NAME "windows"
+  #endif
+|}
+
+
 let bgetenv v =
   let v' =
     try Sys.getenv v |> int_of_string with
@@ -84,6 +105,25 @@ let default_cflags c =
     |> List.filter (fun s -> String.trim s <> "")
   with
   | Not_found ->
+    let os =
+      let header =
+        let file = Filename.temp_file "discover" "os.h" in
+        let fd = open_out file in
+        output_string fd detect_system_header;
+        close_out fd;
+        file
+      in
+      let platform =
+        C.C_define.import c ~includes:[ header ] [ "PLATFORM_NAME", String ]
+      in
+      match List.map snd platform with
+      | [ String "android" ] -> `android
+      | [ String "ios" ]     -> `ios
+      | [ String "linux" ]   -> `linux
+      | [ String "mac" ]     -> `mac
+      | [ String "windows" ] -> `windows
+      | _                    -> `unknown
+    in
     let arch =
       let defines =
         C.C_define.import
@@ -104,10 +144,11 @@ let default_cflags c =
       | _ -> `unknown
     in
     [ (* Basic optimisation *) "-g"; "-O3"; "-Ofast" ]
-    @ (match arch with
-      | `x86_64 -> [ "-march=native"; "-msse2" ]
-      | `arm64  -> [ "-mcpu=apple-m1" ]
-      | _       -> [])
+    @ (match arch, os with
+      | `arm64, `mac -> [ "-mcpu=apple-m1" ]
+      | `arm64, _    -> [ "-march=native" ]
+      | `x86_64, _   -> [ "-march=native"; "-msse2" ]
+      | _            -> [])
     @ [ (* Experimental switches, -ffast-math may break IEEE754 semantics*)
         "-funroll-loops"
       ; "-ffast-math"

--- a/src/owl/config/configure.ml
+++ b/src/owl/config/configure.ml
@@ -84,17 +84,37 @@ let default_cflags =
     |> List.filter (fun s -> String.trim s <> "")
   with
   | Not_found ->
-    [ (* Basic optimisation *)
-      "-g"
-    ; "-O3"
-    ; "-Ofast"
-      (* Experimental switches, -ffast-math may break IEEE754 semantics*)
-    ; "-funroll-loops"
-    ; "-ffast-math"
-    ; (* Configure Mersenne Twister RNG *)
-      "-DSFMT_MEXP=19937"
-    ; "-fno-strict-aliasing"
-    ]
+    let arch =
+      let defines =
+        C.C_define.import
+          c
+          ~includes:[]
+          [ "__x86_64__", Switch
+          ; "__i386__", Switch
+          ; "__aarch64__", Switch
+          ; "__arm__", Switch
+          ]
+      in
+      let open C in
+      match List.map snd defines with
+      | Switch true :: _ -> `x86_64
+      | _ :: Switch true :: _ -> `x86
+      | _ :: _ :: Switch true :: _ -> `arm64
+      | _ :: _ :: _ :: Switch true :: _ -> `arm
+      | _ -> `unknown
+    in
+    [ (* Basic optimisation *) "-g"; "-O3"; "-Ofast" ]
+    @ (match arch with
+      | `x86_64 -> [ "-march=native"; "-msse2" ]
+      | `arm64  -> [ "-mcpu=apple-m1" ]
+      | _       -> [])
+    @ [ (* Experimental switches, -ffast-math may break IEEE754 semantics*)
+        "-funroll-loops"
+      ; "-ffast-math"
+      ; (* Configure Mersenne Twister RNG *)
+        "-DSFMT_MEXP=19937"
+      ; "-fno-strict-aliasing"
+      ]
 
 
 let default_libs = [ "-lm" ]

--- a/src/owl/config/configure.ml
+++ b/src/owl/config/configure.ml
@@ -135,7 +135,6 @@ let default_cflags c =
           ; "__arm__", Switch
           ]
       in
-      let open C in
       match List.map snd defines with
       | Switch true :: _ -> `x86_64
       | _ :: Switch true :: _ -> `x86

--- a/src/owl/config/configure.ml
+++ b/src/owl/config/configure.ml
@@ -87,6 +87,7 @@ let default_cflags =
     [ (* Basic optimisation *)
       "-g"
     ; "-O3"
+    ; "-Ofast"
       (* Experimental switches, -ffast-math may break IEEE754 semantics*)
     ; "-funroll-loops"
     ; "-ffast-math"

--- a/src/owl/config/configure.ml
+++ b/src/owl/config/configure.ml
@@ -146,7 +146,7 @@ let default_cflags c =
     @ (match arch, os with
       | `arm64, `mac -> [ "-mcpu=apple-m1" ]
       | `arm64, _    -> [ "-march=native" ]
-      | `x86_64, _   -> [ "-march=native"; "-msse2" ]
+      | `x86_64, _   -> [ "-march=native"; "-mfpmath=sse"; "-msse2" ]
       | _            -> [])
     @ [ (* Experimental switches, -ffast-math may break IEEE754 semantics*)
         "-funroll-loops"

--- a/src/owl/config/configure.ml
+++ b/src/owl/config/configure.ml
@@ -87,17 +87,11 @@ let default_cflags =
     [ (* Basic optimisation *)
       "-g"
     ; "-O3"
-    ; "-Ofast"
-    ; (* FIXME: experimental switches *)
-      (* "-mavx2"; "-mfma"; "-ffp-contract=fast"; *)
       (* Experimental switches, -ffast-math may break IEEE754 semantics*)
-      "-march=native"
-    ; "-mfpmath=sse"
     ; "-funroll-loops"
     ; "-ffast-math"
     ; (* Configure Mersenne Twister RNG *)
       "-DSFMT_MEXP=19937"
-    ; "-msse2"
     ; "-fno-strict-aliasing"
     ]
 

--- a/src/owl/core/owl_core_utils.c
+++ b/src/owl/core/owl_core_utils.c
@@ -207,7 +207,13 @@ void c_slicing_offset (struct caml_ba_array *X, int64_t *slice, int64_t *offset)
  * Code heavily inspired by Eigen (http://eigen.tuxfamily.org/).
  */
 
-
+#if defined(__arm__) || defined(__aarch64__)
+void query_cache_sizes(int* l1p, int* l2p, int* l3p) {
+  *l1p = 16 * 1024;
+  *l2p = 512 * 1024;
+  *l3p = 512 * 1024;
+}
+#elif
 OWL_INLINE void query_cache_sizes_intel(int* l1p, int* l2p, int* l3p) {
   int cpuinfo[4];
   int l1 = 0, l2 = 0, l3 = 0;
@@ -295,3 +301,4 @@ void query_cache_sizes(int* l1p, int* l2p, int* l3p) {
     *l3p = 512 * 1024;
   }
 }
+#endif

--- a/src/owl/core/owl_macros.h
+++ b/src/owl/core/owl_macros.h
@@ -86,6 +86,8 @@ typedef struct { double r, i; } complex_double;
   #define OWL_ARCH_x86_64 0
 #endif
 
+#if defined(__arm__) || defined(__aarch64__)
+#elif
 #if defined(__PIC__) && OWL_ARCH_i386
   #define CPUID(cpuinfo,func,id) \
     __asm__ __volatile__ ("xchgl %%ebx, %k1; cpuid; xchgl %%ebx,%k1": "=a" (cpuinfo[0]), "=&r" (cpuinfo[1]), "=c" (cpuinfo[2]), "=d" (cpuinfo[3]) : "a" (func), "c" (id));
@@ -96,7 +98,7 @@ typedef struct { double r, i; } complex_double;
   #define CPUID(cpuinfo,func,id) \
     __asm__ __volatile__ ("cpuid": "=a" (cpuinfo[0]), "=b" (cpuinfo[1]), "=c" (cpuinfo[2]), "=d" (cpuinfo[3]) : "0" (func), "2" (id) );
 #endif
-
+#endif
 
 //  Other
 

--- a/test/unit_signal.ml
+++ b/test/unit_signal.ml
@@ -75,7 +75,7 @@ module To_test = struct
     M.set freqz_ref [| 3 |] { Complex.re = -0.0536; im = 0.0464 };
     let f =
       freqz ~n:4 ~whole:false (freqz_num |> M.to_array) (freqz_den |> M.to_array)
-      |> fun (a, b) -> b
+      |> snd
     in
     let max_err =
       (Owl_dense_ndarray.Z.map2 (fun x a -> Complex.sub x a) f freqz_ref

--- a/test/unit_signal.ml
+++ b/test/unit_signal.ml
@@ -74,8 +74,7 @@ module To_test = struct
     M.set freqz_ref [| 2 |] { Complex.re = -0.5; im = -0.5 };
     M.set freqz_ref [| 3 |] { Complex.re = -0.0536; im = 0.0464 };
     let f =
-      freqz ~n:4 ~whole:false (freqz_num |> M.to_array) (freqz_den |> M.to_array)
-      |> snd
+      freqz ~n:4 ~whole:false (freqz_num |> M.to_array) (freqz_den |> M.to_array) |> snd
     in
     let max_err =
       (Owl_dense_ndarray.Z.map2 (fun x a -> Complex.sub x a) f freqz_ref

